[IIIF-373] Configure Terraform to use external module

cachemeoutside opened the pull request:
*Summary of changes*
* Modified terraform configurations to use external modules that point to the following terraform modules:
  * https://github.com/UCLALibrary/aws_terraform_lambda_module
  * https://github.com/UCLALibrary/aws_terraform_s3_module
* Created a new deploy/destroy script to use app.terraform.io for backend state storage
* Introduced storing resources in workspaces on app.terraform.io
* Provided instructions on terraform subpage on how to use app.terraform.io for deploys
* Simplified and organized terraform.tfvars resource file to be more readable

Commit:
Push for workstation switch tomorrow

Commit:
Push deploy script for workstation change

Commit:
Test GPG key

cachemeoutside committed:
Update tfvars source to use new fields. Also reformatted

cachemeoutside committed:
Commit before update

cachemeoutside committed:
merge with master branch

cachemeoutside committed:
get rid of duplicate

cachemeoutside committed:
Fix some missing vars

cachemeoutside committed:
Clean up old files

cachemeoutside committed:
Create instructions and consolidated deploy scirpt

cachemeoutside committed:
Update README to include destroy flag

ksclarke commented on src/main/tools/checkstyle/checkstyle.xml:
```diff
@@ -85,7 +85,7 @@
     </module>
   </module>
   <module name="SuppressionFilter">
-    <property name="file" value="/home/kevin/Workspace/kakadu-lambda-converter/src/main/tools/checkstyle/checkstyle-suppressions.xml" />
```
We need to do something about this. I think we have to have the file checked in for codacy but it's not great to have specific filesystem paths in here. This doesn't have to be solved in this commit, but we probably need a ticket for it.

cachemeoutside committed:
Add function_name as modifiable terraform var

ksclarke reviewed (commented):
I'm going to step through comments/suggestions as I work on testing this out.

1. Create a link from the project README to the terraform README

2. We need to run the backend state file storage script before running `/bin/deploy.sh` (note this in project readme since that's where the deploy.sh is) [EDIT: Oh, if we're replacing the old deploy.sh script we can rename the newly named one to the older]

3. Would be nice to have backend.hcl be a src/main/resources file so we could just put those variables in Maven settings. This would eliminate the need for the *.sample version. This also externalizes the token information so it's not bound to a particular project (or not lost if a dev blows away the project for some reason). Making it a resource would allow us to use actual variables in the file (e.g. ${terraform_username}-) instead of things like "yourtoken". If we don't do that though, it would be good to upcase all the variables to draw attention to them (i.e. is "username-" something I'm supposed to replace? Where as "YOUR_USERNAME_HERE-" is a little more obvious).

4. Add to documentation: To get token for the backend.hcl go to: https://app.terraform.io/app/settings/tokens and generate a new one

5. Don't assume the 'terraform' full path in the `bin/app-terraform-deploy.sh` file. Just have the basic name in the variable and assume they have it on their path? That still gives them the option to put a particular path if they want/need something specific.

6. instead of saying `vi backend.hcl # insert token and organization name into fields` I'd just give a description of how to get these bits of information and leave it up to the user to edit the file to include them (i.e., don't assume vi). This is related to #4 above, too.

7. On the first run, there are errors: https://gist.github.com/ksclarke/1042ac9e8398b016bfa11cacdf6137be -- running it a second time makes them go away so something about the order of the initialization?

(more to come as I work through this)

cachemeoutside committed:
Add backend.hcl to resources

cachemeoutside committed:
moved to be compiled with maven

cachemeoutside committed:
Consolidate deploy script and clean up

cachemeoutside committed:
Update pom to include properties for terraform HCL

cachemeoutside committed:
Rename deploy script to run

cachemeoutside committed:
Add missing variables

cachemeoutside committed:
Add terraform config properties

cachemeoutside committed:
Move README to root

cachemeoutside committed:
Clean up and add run script

cachemeoutside committed:
Test README format

cachemeoutside committed:
Test README format

cachemeoutside committed:
Test README format

cachemeoutside committed:
Testing README

cachemeoutside committed:
Testing README

cachemeoutside committed:
Testing README

cachemeoutside committed:
Testing README

cachemeoutside committed:
Update README

cachemeoutside commented:
@ksclarke this is ready to be reviewed:

1. I've moved the documentation to the root README page
2. backend.hcl is stored in the src directory and is dynamically generated by running maven now
3. see bullet 2
4. This documentation is added to the root README
5. I've reverted to just not using a full path, more of a sysadmin practice, but not practical in a dev workflow
6. I added documentation to the README giving an example of what values to define in the required property fields
7. I fixed this in the new run script. There's a two step process in initializing terraform:
  * Configuring terraform to talk to app.terraform.io by reading the backend.hcl configurations
  * Post-initialize providers(e.g. AWS, Templates) so that terraform can use the provider libs
  * I added a quick comment in the run script describing this

I have to head out in a bit as I'm visiting my sister in the hospital, documentation isn't exactly where I'd want it to be, but let me know if you have more feedback

cachemeoutside commented:
I'm not sure why codacy is complaining. That's odd.

ksclarke commented:
Seems like an internal error. I restarted Codacy's build and it completed successfully.

ksclarke commented:
I've made some suggestions on https://github.com/UCLALibrary/kakadu-lambda-converter/pull/20

This PR is good to go after you've reviewed those and accepted/rejected them.

ksclarke committed:
IIIF-373 suggestions (#20)

* Use prefix to provide default function name
* Do the terraform apply by default
* Remove .sh from check_plan for consistency
* Update README documentation

cachemeoutside-bot committed:
Remove incremental deploy stanzas, deprecated

ksclarke reviewed (approved):
Looks good!
